### DISCUSSION TITLE
Does Fooocus support Wildcards?

### DISCUSSION THREAD

**Discussion by kirbykirby** [lllyasviel/Fooocus#494] [Q&A] [answered]
Does Fooocus support Wildcards like the Dynamic Prompts extension in A1111's WebUI?

**Comment by barepixels** [comment, marked as answer]
https://github.com/MoonRide303/Fooocus-MRE and 
https://github.com/runew0lf/RuinedFooocus does


Wildcards are now supported place see wildcards\colors.txt for an example. In your prompt make sure you type
 `__<filename>__` 
to activate ie 
`shiny new __colors__ Chevrolet pickup truck with big wheels`

**Reply by ghost** [reply]
There is no "auto-complete" feature. So, it can kind of work if you have 1 or max 2 cards

**Comment by agarzon** [comment]
I which this to be implemented here 😪

**Reply by agarzon** [reply]
Actually, my bad. I just check and see the wildcard folder so I guess it is already supported ! 😀

**Comment by MoonRide303** [comment]
@kirbykirby @agarzon I added support for wildcards 3 days ago (in v2.1.687, https://github.com/lllyasviel/Fooocus/pull/712), and also improved it a bit next day (in v2.1.696, https://github.com/lllyasviel/Fooocus/pull/722) - using same wildcard multiple times can now yield different values (`__color__ cat in a __color__ hat` can now pick 2 different colors).

**Reply by agarzon** [reply]
I noticed the wildcard value is not reflected in the log file

**Reply by MoonRide303** [reply]
@agarzon Try current version :).

**Reply by agarzon** [reply]
it works 😃

**Comment by aolko** [comment]
Do wildcards support nesting? I.e. calling wildcards in wildcards

**Reply by MoonRide303** [reply]
@aolko Nope, it replaces only wildcards used directly in the original prompt.

**Reply by aolko** [reply]
> @aolko Nope, it replaces only wildcards used directly in the original prompt.

@MoonRide303 but what if i want to reference

```
colors.txt
red,blue,green,yellow,white,black
```
```
brands.txt
bmw,audi,jeep,ferrari
```
```
car.txt
a __color__ __brand__ sports car
```

```
__car__ on the race track
```

**Reply by MoonRide303** [reply]
@aolko You would need to directly reference those wildcards, like `a __color__ __brand__ sports car on the race track` (assuming properly defined brand.txt is available as a wildcard - mind choices need to be put in separate lines, not in comma-separated list).

**Reply by lllyasviel** [reply]
interesting. I spent 10 minutes to add a bfs for this trick

**Reply by 925-Studio** [reply]
Any chance to support `.yaml` format wildcard like Dynamic Prompt extension does? Thanks in advance.

**Reply by aolko** [reply]
Don't forget `.toml`/`.ini`

**Comment by R4290N** [comment]
> Не забывайте `.toml`/`.ini`

what do you mean by that?

**Comment by GrafKnusprig** [comment]
Is it possible to use wildcards directly in prompt and not in a file?
For example "a girl in the {rain|sun|snow}".
I find it a bit complicated to have to create a file every time for such small things.
Thanks

**Reply by mashb1t** [reply]
@GrafKnusprig only after https://github.com/lllyasviel/Fooocus/pull/1503 has been approved and merged.

**Reply by metercai** [reply]
Enhanced on the basis of [#1503](https://github.com/lllyasviel/Fooocus/pull/1503) and adds the ability to group words with wildcards in SimpleSDXL  derived from Fooocus : https://github.com/metercai/SimpleSDXL
For example `a beautiful [man, woman] wearing a [__color__:2] [__flower__:L2:4] suit`
There are three modes in this example:
1, `[man, woman]`:  embedded in prompt
2, `[__color__:2]`: randomly select from wildcards
3, `[__flower__:L2:4]`: select in order of specified positions from wildcards
and automatically recognize and trigger auxiliary input tools, such as:
<img width="600" src="https://github.com/lllyasviel/Fooocus/assets/5652458/114d018f-397c-45e4-9164-97d4bd38da70">

**Comment by Ckrotky** [comment]
@mashb1t, Hi 😀
Can I specify a list of ready-made prompts (rather than single values) in wildcard to get random results with different compositions (of course each prompt is written on a single line and each new prompt is written on a new line)?


For example, 
/content/Fooocus/wildcards/prompts.txt:
Woman in a cafe drinking coffee 
A woman is working on her laptop
Woman taking selfies near a fountain
etc.

Main Prompt:
`HDR, Portrait, __prompts__`

**Reply by ILikeToasters** [reply]
Yes that works.

**Reply by mashb1t** [reply]
yes, you can do that

**Reply by Ckrotky** [reply]
@ILikeToasters, @mashb1t 
Thank you for the answer🙌
